trace-normalization/src/normalize_utils.rs

```diff
@@ -25,23 +32,141 @@ pub(crate) fn truncate_utf8(s: &str, limit: usize) -> &str {
+    // Fast path: Check if the tag is valid and only contains ASCII characters,
+    // if yes return it as-is right away. For most use-cases this reduces CPU usage.
+    if is_normalized_ascii_tag(tag) {
+        return Ok(tag.to_string());
+    }
+
+    anyhow::ensure!(!tag.is_empty(), "Normalizer Error: Empty tag name.");
+
+    // given a dummy value
+    let mut last_char: char = 'a';
+
+    let mut result = String::with_capacity(tag.len());
+
+    let char_vec: Vec<char> = tag.chars().collect();
+
+    for cur_char in char_vec {
+        if result.len() == MAX_TAG_LEN {
+            break;
+        }
+        if cur_char.is_lowercase() {
+            result.push(cur_char);
+            last_char = cur_char;
+            continue;
+        }
+        if cur_char.is_uppercase() {
+            let mut iter = cur_char.to_lowercase();
+            if let Some(c) = iter.next() {
+                result.push(c);
+                last_char = c;
+            }
+            continue;
+        }
+        if cur_char.is_alphabetic() {
+            result.push(cur_char);
+            last_char = cur_char;
+            continue;
+        }
+        if cur_char == ':' {
+            result.push(cur_char);
+            last_char = cur_char;
+            continue;
+        }
+        if !result.is_empty()
+            && (cur_char.is_ascii_digit() || cur_char == '.' || cur_char == '/' || cur_char == '-')
+        {
+            result.push(cur_char);
+            last_char = cur_char;
+            continue;
+        }
+        if !result.is_empty() && last_char != '_' {
+            result.push('_');
+            last_char = '_';
+        }
+    }
+
+    if last_char == '_' {
+        result.remove(result.len() - 1);
+    }
+
+    Ok(result.to_string())
+}
+
+pub(crate) fn is_normalized_ascii_tag(tag: &str) -> bool {
+    if tag.is_empty() {
+        return true;
+    }
+    if tag.len() > MAX_TAG_LEN {
+        return false;
+    }
+    match tag.chars().next() {
+        Some(c) => {
+            if !is_valid_ascii_start_char(c) {
+                return false;
+            }
+        }
+        None => return false,
+    }
+
+    for mut i in 0..tag.len() {
+        let cur_char: char = match tag.chars().nth(i) {
+            Some(c) => c,
+            None => return false,
+        };
+        if is_valid_ascii_tag_char(cur_char) {
+            continue;
+        }
+        if cur_char == '_' {
+            // an underscore is only okay if followed by a valid non-underscore character
+            i += 1;
+            match tag.chars().nth(i) {
+                Some(c) => {
+                    if !is_valid_ascii_tag_char(c) {
+                        return false;
+                    }
+                }
+                None => return false,
+            };
+        } else {
+            return false;
+        }
+    }
+    true
+}
+
+pub(crate) fn is_valid_ascii_start_char(c: char) -> bool {
+    ('a'..='z').contains(&c) || c == ':'
```

Comment:
Same deal here, also, are only lower case characters valid as a start character?

Reply:
yes, thats correct ([corresponding function in the go agent normalizer](https://github.com/DataDog/datadog-agent/blob/main/pkg/trace/traceutil/normalize.go))